openvm/src/empirical_constraints.rs

```diff
@@ -0,0 +1,386 @@
+use itertools::Itertools;
+use openvm_circuit::arch::VmCircuitConfig;
+use openvm_sdk::StdIn;
+use openvm_stark_backend::p3_maybe_rayon::prelude::IntoParallelIterator;
+use openvm_stark_backend::p3_maybe_rayon::prelude::ParallelIterator;
+use openvm_stark_sdk::openvm_stark_backend::p3_field::PrimeField32;
+use powdr_autoprecompiles::empirical_constraints::{
+    intersect_partitions, DebugInfo, EmpiricalConstraints,
+};
+use powdr_autoprecompiles::DegreeBound;
+use std::collections::btree_map::Entry;
+use std::collections::HashMap;
+use std::collections::{BTreeMap, BTreeSet};
+
+use crate::trace_generation::do_with_trace;
+use crate::{CompiledProgram, OriginalCompiledProgram};
+
+#[derive(Debug)]
+struct Row {
+    pc: u32,
+    timestamp: (u32, u32),
```

Comment:
Why is timestamp a tuple?

Reply:
segment id, timestamp value within this segment. I made this clear in #3474